work.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"context"
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/heptio/workgroup"
+	"go.elastic.co/apm"
+)
+
+type worker struct {
+	*apmLogger
+	*apm.Tracer
+	runTimeout time.Duration
+
+	// not to be modified concurrently
+	workgroup.Group
+}
+
+type Report struct {
+	Count int
+	Start time.Time
+	End   time.Time
+}
+
+type sampler struct {
+	count int
+}
+
+func (s *sampler) Sample(apm.TraceContext) bool {
+	s.count++
+	return true
+}
+
+func (w *worker) Work() (Report, error) {
+	if w.runTimeout > 0 {
+		w.Add(func(done <-chan struct{}) error {
+			select {
+			case <-done:
+				return nil
+			case <-time.After(w.runTimeout):
+				return nil // time expired
+			}
+		})
+	}
+
+	var s sampler
+	w.Tracer.SetSampler(&s)
+
+	report := Report{}
+	report.Start = time.Now()
+	err := w.Run()
+	report.End = time.Now()
+	report.Count = s.count
+	return report, err
+}
+
+func (w *worker) addTransactions(limit, spanMin, spanMax int) *worker {
+	if limit <= 0 {
+		return w
+	}
+	generateSpan := func(ctx context.Context) {
+		time.Sleep(time.Duration(rand.Intn(3)) * time.Millisecond)
```

Comment:
Since we intend to use this for benchmarking, I'd remove as many nondeterminism sources as possible, like the `rand` calls here. This also caps workload massively iiuc, I have not been able to push more than 100 transactions in 30 seconds due to the `time.Sleep` calls in this function.

If you just want to show some span cascading in the UI, I think this should suffice:
```go
		span, ctx := apm.StartSpanOptions(ctx, "I'm a span", "gen.era.ted",
			apm.SpanOptions{Start:time.Now().Add(-time.Duration(rand.Intn(500)))})

```

Reply:
That's fair.  I do think we need to replace it with some sort of throttle - events per second or similar.  We'll have to adjust span frames min duration accordingly.

Reply:
ok, what do you think then about one `time.Sleep(time.Millisecond)` call in L66 and remove L71, for now?

Reply:
Inspired by your comment IRL, I removed the sleep altogether and set the min span duration threshold to 1ns, looks good to me:

![image](https://user-images.githubusercontent.com/83483/55657048-0249a500-57c7-11e9-8b5c-8ac35c6a3779.png)